File: go/ontology/registry_test.go
// SPDX-License-Identifier: Apache-2.0
package ontology_test

import (
	"context"
	"testing"

	"github.com/jeffs-brain/memory/go/ontology"
	"github.com/jeffs-brain/memory/go/store/mem"
)

func newTestRegistry(t *testing.T) *ontology.Registry {
	t.Helper()
	ms := mem.New()
	t.Cleanup(func() { _ = ms.Close() })
	cfg := ontology.FileOntologyStoreConfig{
		BrainID:   "brain-1",
		ProjectID: "proj-1",
		OrgID:     "org-1",
	}
	store, err := ontology.NewFileOntologyStore(ms, cfg)
	if err != nil {
		t.Fatalf("NewFileOntologyStore: %v", err)
	}
	return ontology.NewRegistry(store)
}

func TestRegistryRegisterType(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	reg := newTestRegistry(t)

	def := ontology.TypeDefinition{
		Type:        "entity.invoice",
		Label:       "Invoice",
		Description: "Financial invoice",
		CreatedAt:   "2026-01-01T00:00:00Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := reg.RegisterType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("RegisterType: %v", err)
	}

	resolved, err := reg.Resolve(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}
	found := false
	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.invoice" && rt.Label == "Invoice" {
			found = true
			break
		}
	}
	if !found {
		t.Fatal("expected entity.invoice in resolved ontology after RegisterType")
	}
}

func TestRegistryDeprecateType(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	reg := newTestRegistry(t)

	def := ontology.TypeDefinition{
		Type:        "entity.invoice",
		Label:       "Invoice",
		Description: "Financial invoice",
		CreatedAt:   "2026-01-01T00:00:00Z",
		Status:      ontology.TypeStatusActive,
	}
	if err := reg.RegisterType(ctx, ontology.ScopeBrain, def); err != nil {
		t.Fatalf("RegisterType: %v", err)
	}
	if err := reg.DeprecateType(ctx, ontology.ScopeBrain, "entity.invoice"); err != nil {
		t.Fatalf("DeprecateType: %v", err)
	}

	resolved, err := reg.Resolve(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}
	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.invoice" {
			t.Fatal("deprecated type should not appear in resolved ontology")
		}
	}
}

func TestRegistryProposeType_CreatesNew(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	reg := newTestRegistry(t)

	if err := reg.ProposeType(ctx, ontology.ScopeBrain, "nodeType", "entity.invoice", "Discovered in procurement docs"); err != nil {
		t.Fatalf("ProposeType: %v", err)
	}

	types, err := reg.Resolve(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}
	// Proposed types should NOT appear in resolved (only active types resolve)
	for _, rt := range types.NodeTypes {
		if rt.Type == "entity.invoice" && rt.Scope == "brain" {
			// The type exists but should be proposed status, which is filtered
			t.Fatal("proposed type should not appear as active in resolved ontology")
		}
	}
}

func TestRegistryProposeType_SkipsSimilar(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	reg := newTestRegistry(t)

	// "entity.customer" is built-in with label "Customer (Entity)".
	// Proposing "entity.customers" should be skipped because
	// "Customers (Entity)" is very similar to "Customer (Entity)".
	if err := reg.ProposeType(ctx, ontology.ScopeBrain, "nodeType", "entity.customers", "Should be skipped"); err != nil {
		t.Fatalf("ProposeType: %v", err)
	}

	// Verify the type was NOT created (it was skipped as duplicate)
	resolved, err := reg.Resolve(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}
	for _, rt := range resolved.NodeTypes {
		if rt.Type == "entity.customers" {
			t.Fatal("similar type should have been skipped")
		}
	}
}

func TestRegistryProposeType_InvalidCategory(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	reg := newTestRegistry(t)

	err := reg.ProposeType(ctx, ontology.ScopeBrain, "invalid", "entity.test", "reason")
	if err == nil {
		t.Fatal("expected error for invalid category")
	}
}

func TestRegistryResolve_BuiltInOnly(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	reg := newTestRegistry(t)

	resolved, err := reg.Resolve(ctx, "brain-1", "proj-1", "org-1")
	if err != nil {
		t.Fatalf("Resolve: %v", err)
	}
	if len(resolved.NodeTypes) != 30 {
		t.Fatalf("expected 30 node types, got %d", len(resolved.NodeTypes))
	}
	if len(resolved.EdgeTypes) != 29 {
		t.Fatalf("expected 29 edge types, got %d", len(resolved.EdgeTypes))
	}
	if len(resolved.BusinessCategories) != 8 {
		t.Fatalf("expected 8 business categories, got %d", len(resolved.BusinessCategories))
	}
}

File: sdks/ts/memory/src/ontology/registry.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it, beforeEach } from 'vitest'
import { createMemStore } from '../store/memstore.js'
import type { OntologyTypeDefinition } from './types.js'
import { createFileOntologyStore, type OntologyStore } from './store.js'
import { Registry } from './registry.js'

function makeRegistry(): Registry {
  const backingStore = createMemStore()
  const ontologyStore = createFileOntologyStore(backingStore, {
    brainId: 'brain-1',
    projectId: 'proj-1',
    orgId: 'org-1',
  })
  return new Registry({ store: ontologyStore })
}

describe('Registry', () => {
  let registry: Registry

  beforeEach(() => {
    registry = makeRegistry()
  })

  describe('registerType', () => {
    it('adds a type to the store', async () => {
      const def: OntologyTypeDefinition = {
        type: 'entity.invoice',
        label: 'Invoice',
        description: 'Financial invoice',
        createdAt: '2026-01-01T00:00:00Z',
        status: 'active',
      }
      await registry.registerType('brain', def)

      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      const invoice = resolved.nodeTypes.find((t) => t.type === 'entity.invoice')
      expect(invoice).toBeDefined()
      expect(invoice!.label).toBe('Invoice')
    })
  })

  describe('deprecateType', () => {
    it('marks a type as deprecated', async () => {
      const def: OntologyTypeDefinition = {
        type: 'entity.invoice',
        label: 'Invoice',
        description: 'Financial invoice',
        createdAt: '2026-01-01T00:00:00Z',
        status: 'active',
      }
      await registry.registerType('brain', def)
      await registry.deprecateType('brain', 'entity.invoice')

      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      const invoice = resolved.nodeTypes.find((t) => t.type === 'entity.invoice')
      expect(invoice).toBeUndefined()
    })

    it('throws for nonexistent type', async () => {
      await expect(
        registry.deprecateType('brain', 'entity.missing'),
      ).rejects.toThrow('not found')
    })
  })

  describe('proposeType', () => {
    it('creates a proposed type for genuinely new types', async () => {
      await registry.proposeType('brain', 'nodeType', 'entity.invoice', 'Discovered in procurement docs')

      // Proposed types should NOT appear in resolved (only active resolve)
      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      const invoice = resolved.nodeTypes.find((t) => t.type === 'entity.invoice' && t.scope === 'brain')
      expect(invoice).toBeUndefined()
    })

    it('skips when similar type already exists', async () => {
      // "entity.customer" is built-in. Proposing "entity.customers"
      // should be skipped since "Customers (Entity)" is very similar
      // to "Customer (Entity)".
      await registry.proposeType('brain', 'nodeType', 'entity.customers', 'Should be skipped')

      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      const customers = resolved.nodeTypes.find((t) => t.type === 'entity.customers')
      expect(customers).toBeUndefined()
    })

    it('creates edge type proposals', async () => {
      await registry.proposeType('brain', 'edgeType', 'ships_to', 'Logistics relation')

      // ships_to is not a built-in edge type, so it should be created as proposed
      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      // It's proposed so won't appear in resolved (active only)
      const shipsTo = resolved.edgeTypes.find((t) => t.type === 'ships_to' && t.scope === 'brain')
      expect(shipsTo).toBeUndefined()
    })
  })

  describe('resolve', () => {
    it('returns all built-in types when store is empty', async () => {
      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      expect(resolved.nodeTypes).toHaveLength(30)
      expect(resolved.edgeTypes).toHaveLength(29)
      expect(resolved.businessCategories).toHaveLength(8)
    })

    it('includes registered types in resolution', async () => {
      const def: OntologyTypeDefinition = {
        type: 'entity.invoice',
        label: 'Invoice',
        description: 'Financial invoice',
        createdAt: '2026-01-01T00:00:00Z',
        status: 'active',
      }
      await registry.registerType('brain', def)

      const resolved = await registry.resolve('brain-1', 'proj-1', 'org-1')
      expect(resolved.nodeTypes).toHaveLength(31)
    })
  })
})
